### PR TITLE
perf(filters): default box blur to separable convolutions

### LINKED ISSUE #2246: Improve speed for box blur (mean filter)
## 🚀 Feature

Originally from this: https://github.com/kornia/kornia/issues/720#issuecomment-881859300

In theory box blur (mean filter) can be implemented in O(n) time.

Two solutions

- Use 2 separable 1D kernels (like `gaussian_blur2d()`) -> O(nk) time, should be simple. Only launch 2 pytorch calls.
- Use intergral image (like [here](https://github.com/wuhuikai/DeepGuidedFilter/blob/master/GuidedFilteringLayer/GuidedFilter_PyTorch/guided_filter_pytorch/box_filter.py)) -> O(n) time. It seems like there will be quite a lot of overhead for small kernels

## Additional context

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by ducha-aiki**
## What does this pull request do?

Changes `box_blur` and `BoxBlur` to default to `separable=True`: two 1D convolutions instead of one 2D convolution. This proposal targets the substantial CPU small-batch and large-kernel wins, while keeping `separable=False` available.

**Draft for a default-policy decision:** this is not a universal speed-up. Small-kernel MPS workloads and some large CPU batches regress. The measurements below disclose those regressions; CUDA was unavailable and needs measurement before deciding to merge.

- Preserve the flagship benchmark's historical `box_blur` row with explicit `separable=False`, and add `box_blur_separable` with explicit `True` (including opt-in compiled timing).
- Cover default selection, odd/even/rectangular/unit kernels, all four border modes, input gradients, both module checkpoint paths, and both gradcheck paths. Preserve explicit nonseparable export coverage.
- Correct `RandomBoxBlur` documentation: its explicit `normalized=True`/`separable=True` now agrees with the function default; its behavior is unchanged. Guided blur also passes an explicit setting and is unchanged.
- Clone the module's expanded kernel buffers so `load_state_dict` can write to them without overlapping-memory errors.

**Compatibility:** outputs are mathematically equivalent for ordinary finite inputs, but not bit-identical. `BoxBlur` default state dictionaries change from `kernel` to `kernel_x`/`kernel_y`. Construct `BoxBlur(..., separable=False)` before loading an old default checkpoint. This is documented in the API and breaking-change fragment; checkpoints are not silently converted.

## Numerical comparison

Compared explicit False/True on CPU and MPS, with seed 1; shape `(2,3,20,22)`; kernels `1,3,5,15,(2,4),(3,7)`; constant/reflect/replicate/circular borders; uniform `[0,1]`, signed `[-1,1]`, constant-one, and impulse inputs. Gradients use fixed-seed random upstream weights. Table gives maxima across the cases; relative L2 is `||new-old|| / ||old||`, not elementwise relative error near zero. This is sampled evidence, not a bound for arbitrary inputs.

| Device/dtype | Max output absolute difference | Max output relative L2 | Max gradient absolute difference | Max gradient relative L2 |
| --- | ---: | ---: | ---: | ---: |
| CPU float32 | 7.75e-7 | 5.50e-7 | 2.38e-7 | 2.86e-7 |
| CPU float64 | 2.33e-15 | 1.78e-15 | 1.33e-15 | 4.33e-16 |
| CPU float16 | 9.77e-4 | 7.64e-4 | 2.93e-3 | 2.58e-3 |
| CPU bfloat16 | 7.81e-3 | 6.85e-3 | 2.34e-2 | 1.93e-2 |
| MPS float32 | 2.26e-6 | 2.26e-6 | 2.98e-7 | 2.81e-7 |
| MPS float16 | 9.77e-4 | 7.78e-4 | 1.95e-3 | 8.36e-4 |
| MPS bfloat16 | 7.81e-3 | 6.85e-3 | 1.56e-2 | 5.90e-3 |

Float32/float64 differences are small. Half-precision rounding is more visible: CPU bfloat16 gradients reach ~1.9% relative L2 difference. The half-precision gradient comparison test allows two dtype epsilons in both absolute and relative tolerance to account for the separately rounded passes and repeated-border contributions. The explicit False path remains available when preserving previous rounding is necessary.

## Performance comparison

Apple M1, macOS, Python 3.11.14, PyTorch 2.14.0; float32 `(B,3,256,256)`, reflect padding, eager public `box_blur` with no `separable` argument. Base `48044c6ba` versus this branch, each run from its own checkout using stdin and the same explicit Python interpreter; verified `kornia.__file__` in each run. Seed 1, `benchmarks.common.time_us(min_run_time=1.0)` (warmup + blocked-autorange, **one timed CPU thread**), MPS synchronization inside the timed region. Timings are directional on a shared desktop; MPS IQRs in particular are large. Lower time is better; speed-up below 1 means a regression.

| Device | Batch | Kernel | Base median / IQR (ms) | New median / IQR (ms) | Speed-up |
| --- | ---: | ---: | ---: | ---: | ---: |
| CPU | 1 | 3 | 0.619 / 0.018 | 0.400 / 0.028 | 1.55x |
| CPU | 1 | 5 | 1.839 / 0.108 | 0.576 / 0.021 | 3.19x |
| CPU | 1 | 15 | 35.365 / 0.957 | 2.346 / 0.124 | 15.07x |
| CPU | 8 | 3 | 9.952 / 0.889 | 9.188 / 0.427 | 1.08x |
| CPU | 8 | 5 | 31.719 / 1.847 | 12.452 / 1.109 | 2.55x |
| CPU | 8 | 15 | 314.859 / 15.081 | 32.085 / 1.719 | 9.81x |
| CPU | 32 | 3 | 30.927 / 3.905 | 51.859 / 2.834 | 0.60x |
| CPU | 32 | 5 | 32.443 / 3.089 | 60.618 / 4.378 | 0.54x |
| CPU | 32 | 15 | 611.804 / 2.106 | 206.230 / 12.033 | 2.97x |
| MPS | 1 | 3 | 0.919 / 0.123 | 1.396 / 0.244 | 0.66x |
| MPS | 1 | 5 | 1.004 / 0.225 | 1.488 / 0.359 | 0.67x |
| MPS | 1 | 15 | 2.262 / 0.090 | 1.492 / 0.633 | 1.52x |
| MPS | 8 | 3 | 1.791 / 0.193 | 3.443 / 1.397 | 0.52x |
| MPS | 8 | 5 | 3.464 / 0.167 | 3.884 / 0.121 | 0.89x |
| MPS | 8 | 15 | 7.515 / 2.148 | 4.282 / 1.305 | 1.75x |
| MPS | 32 | 3 | 3.621 / 1.012 | 5.746 / 0.970 | 0.63x |
| MPS | 32 | 5 | 4.165 / 0.934 | 7.955 / 2.010 | 0.52x |
| MPS | 32 | 15 | 9.677 / 0.091 | 8.685 / 1.591 | 1.11x |

On this CPU, 5x5 improves 3.19x at batch 1 and 2.55x at batch 8, but regresses to 0.54x at batch 32. Large 15x15 kernels improve 2.97–15.07x on CPU. Small MPS kernels regress. These losses are why the PR is a draft.

<details><summary>Timing reproduction (run via stdin from each checkout root)</summary>

Use the same explicit environment interpreter in each checkout; printing the import path guards against accidentally measuring the primary editable checkout twice.

```python
import itertools
import os
import sys
import torch
import kornia
from benchmarks.common import run_metadata, save_json, time_us
print('kornia:', kornia.__file__, 'python:', sys.executable, flush=True)
torch.set_num_threads(4)
for device in ['cpu', 'mps']:
    rows = []
    for batch, kernel in itertools.product((1, 8, 32), (3, 5, 15)):
        torch.manual_seed(1)
        x = torch.rand(batch, 3, 256, 256, device=device)
        sync = torch.mps.synchronize if device == 'mps' else None
        median, iqr = time_us(lambda: kornia.filters.box_blur(x, kernel), min_run_time=1.0, sync=sync)
        rows.append(dict(op='box_blur', backend='kornia (eager)', batch=batch, kernel=kernel,
                         height=256, width=256, dtype='float32', median_us=median, iqr_us=iqr,
                         throughput_per_s=batch * 1e6 / median))
        print(device, batch, kernel, round(median, 2), round(iqr, 2), flush=True)
    # Optionally save rows with common.save_json and run_metadata(torch.device(device)).

```
</details>

## What did you check?

- `pixi run test-module tests/filters/test_blur.py tests/augmentation/test_augmentation.py::TestRandomBoxBlur --device=cpu --dtype=float32,float64,float16,bfloat16`: **319 passed**.
- `pixi run test-module tests/filters/test_blur.py --device=mps --dtype=float32,float16,bfloat16`: **220 passed, 6 skipped** (float64 gradcheck is unsupported on MPS).
- `KORNIA_TEST_OPTIMIZER=inductor pixi run test-module tests/filters/test_blur.py -k dynamo --device=cpu --dtype=float32`: **8 passed**.
- Box-blur augmentation convention selection: **7 passed**.
- Manual default and explicit-nonseparable `torch.export`, ONNX graph validation, and ONNX Runtime output parity: **passed**.
- `pixi run pre-commit-all`: **passed**.
- Independent review plus CPU float32/float64/float16/bfloat16 filter-test reruns: no actionable findings.
- CUDA, older supported PyTorch versions, and end-to-end training were not tested. The full export-support snapshot was not regenerated; its existing cases are pinned to preserve the previously measured paths.

## Related issue or discussion

Follow-up to the separable implementation in #2328 (which closed #2246). No new algorithm or runtime dependency.

## How did AI help?

Codex implemented the change, measured numerical and timing differences, ran validation, and performed an independent code review. This draft explicitly leaves the default-policy tradeoff for maintainer review.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6-astra).

### CODE REVIEW

**Comment by ducha-aiki** [comment]
Review of `3382b96e081f66896ba6d2337ede4c69612b1673` — **intended verdict: APPROVE** (the authenticated reviewer is also the author; GitHub does not permit formal self-review).

I found no actionable correctness issue in the default change, state-dict fix, tests, documentation, or benchmark split. The focused box-blur suite passed across CPU float16, bfloat16, float32, and float64 (299 passed); the merge tree against current `main` is clean and all 45 reported checks are green.

This remains a draft for the documented default-policy decision. It also overlaps with #4647, which includes the same default change; only one of the two should land.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).
